calico_containers/tests/st/utils/workload.py

```diff
@@ -125,15 +128,44 @@ def assert_can_ping(self, ip, retries=0):
         command = ' '.join(args)
 
         ping = partial(self.execute, command)
-        return retry_until_success(ping,
+        return ping
+
+    def assert_can_ping(self, ip, retries=0):
+        """
+        Execute a ping from this workload to the ip. Assert than a workload
+        can ping an IP. Use retries to compensate for network uncertainty and
+        convergence.
+
+        :param ip:  The IP address (str or IPAddress) to ping.
+        :param retries: The number of retries.
+        :return: None.
+        """
+        return retry_until_success(self._get_ping_function(ip),
                                    retries=retries,
-                                   ex_class=CalledProcessError)
+                                   ex_class=CommandExecError)
 
     def assert_cant_ping(self, ip, retries=0):
-        for retry in range(retries + 1):
-            self.assert_can_ping(ip)
-            if retry >= retries:
-                raise Exception("Workload can unexpectedly ping %s" % ip)
+        """
+        Execute a ping from this workload to the ip.  Assert that the workload
+        cannot ping an IP.  Use retries to compensate for network uncertainty
+        and convergence.
+
+        :param ip:  The IP address (str or IPAddress) to ping.
+        :param retries: The number of retries.
+        :return: None.
+        """
+        ping = self._get_ping_function(ip)
+        def cant_ping():
+            try:
+                ping()
+            except CommandExecError:
+                pass
+            else:
+                raise Exception("%s can ping %s" % (self, ip))
+
+        return retry_until_success(cant_ping,
+                                   retries=retries,
+                                   ex_class=Exception)
```

Comment:
Yes I can create a specific exception for this - the reason I didn't was simplicity - it wasn't correct to raise a CommandExecError and I felt that if we have a proper exception for this then we probably should for the assert_can_ping.

I can update.